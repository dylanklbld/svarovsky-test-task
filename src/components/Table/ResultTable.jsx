import '../../styles/table.css'

import React, { useEffect, useState } from 'react'

import PropTypes from 'prop-types'
import ScrollToTopButton from '../ScrollToButton'
import useEffectExceptMount from '../../hooks/useEffectExceptMount'
import { useIsInView } from '../../hooks/useIsInVIew'

const defaultOptions = {
    perPage: 16
}

export const ResultTableWrapper = ({ title, handleFetchDataChunck, options = defaultOptions }) => {
    const [chunckData, setChunckData] = useState(null)
    const [fullData, setFullData] = useState([])
    const [page, setPage] = useState(1)
    const [isBusy, setIsBusy] = useState(false)

    const [totalPages, setTotalPages] = useState(1)

    useEffect(() => {
        handleFetchDataChunck(setChunckData)
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [])

    useEffectExceptMount(() => {
        if (chunckData && chunckData.results) {
            const results = chunckData.results[0]
            setTotalPages(results.nbPages)
            setFullData(fullData.concat(results.hits))
        }
    }, [chunckData])


    return fullData && <ResultTableComponent {...{ title, isBusy }}
        data={fullData}
        renderLoadingButton={() => {
            return page < totalPages ? <button disabled={isBusy} onClick={async () => {
                setIsBusy(true)
                setPage(page + 1)
                await handleFetchDataChunck(setChunckData, Object.assign({ page: page + 1 }, options))
                setIsBusy(false)
            }}>Load more...</button> : <div>Done</div>
        }} />
}

ResultTableWrapper.propTypes = {
    title: PropTypes.string,
    handleFetchDataChunck: PropTypes.func.isRequired
}

export const ResultTableComponent = ({ title, data, renderLoadingButton, isBusy = false }) => {
    const [renderToTopButton, setRenderToTopButton] = useState(false)
    const [refTopDiv, inView] = useIsInView()

    React.useEffect(() => {
        setRenderToTopButton(!inView)
    }, [inView])

    return data ? <React.Fragment>
        <div ref={refTopDiv} />
        <table>
            <caption>{title}</caption>
            <thead>
                <tr>
                    <th scope="col">Product</th>
                    <th scope="col">Overview</th>
                    <th scope="col">Price</th>
                    <th scope="col">Rating</th>
                </tr>
            </thead>
            <tbody>
                {data.map((value) => {
                    const productId = value ? value['ikea_id'] : 0
                    return <tr key={productId}>
                        <td data-label="Product">

                            <div className="container">
                                <div className="product"><img
                                    className="fit-picture"
                                    src={`${value.image}`}
                                    alt={value['name']} /></div>
                            </div>

                        </td>
                        <td data-label="Overview">{value.description}</td>
                        <td data-label="Price">{value.price}</td>
                        <td data-label="Rating">{value.rating}</td>
                    </tr>
                })}
            </tbody>
        </table>
        {isBusy && <div>BUSY STATE</div>}
        {renderLoadingButton && renderLoadingButton()}
        {renderToTopButton && <ScrollToTopButton />}

    </React.Fragment> : <div>"No data"</div>
}

ResultTableComponent.propTypes = {
    title: PropTypes.string.isRequired,
    data: PropTypes.array.isRequired,
    handleFetchDataChunck: PropTypes.func.isRequired
}
