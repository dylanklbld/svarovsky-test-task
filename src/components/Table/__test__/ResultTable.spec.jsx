import "@testing-library/jest-dom"

import  { MULTI_PAGE_DATA, ONE_PAGE_DATA } from '../../../mockdata/tableData';
import { fireEvent, render, wait } from '@testing-library/react'

import React from 'react'
import { ResultTableWrapper } from '../ResultTable'

it('should render empty', () => {
    const { container } = render(<ResultTableWrapper title={'Empty'} handleFetchDataChunck={()=>([])} />)

    expect(container).toMatchSnapshot()
})

it('should render one page table', async () => {
    const { container, getByText } = render(<ResultTableWrapper title={'One page'} handleFetchDataChunck={(stateUpdate)=>(stateUpdate(ONE_PAGE_DATA))} />)

    await wait()

    // checking for loading button
    getByText('Done')

    expect(container).toMatchSnapshot()
})

it('should render multi page table', async () => {
    const pageDataCheckingSpy = jest.fn()
    const { getByText } = render(<ResultTableWrapper title={'Multi page'} 
    
    handleFetchDataChunck={(stateUpdate, options)=>{
        stateUpdate(MULTI_PAGE_DATA)
        options && pageDataCheckingSpy(options)
    }} />)

    await wait()
    fireEvent.click(getByText('Load more...'))
    
    await wait()
    expect(pageDataCheckingSpy).toHaveBeenCalledWith({"page": 2, "perPage":16})
})

it('should render multi page table', () => {
    const { container } = render(<ResultTableWrapper title={'Empty'} handleFetchDataChunck={()=>([])} />)

    expect(container).toMatchSnapshot()
})