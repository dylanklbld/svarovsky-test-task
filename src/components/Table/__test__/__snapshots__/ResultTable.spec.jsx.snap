// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render empty 1`] = `
<div>
  <div />
  <table>
    <caption>
      Empty
    </caption>
    <thead>
      <tr>
        <th
          scope="col"
        >
          Product
        </th>
        <th
          scope="col"
        >
          Overview
        </th>
        <th
          scope="col"
        >
          Price
        </th>
        <th
          scope="col"
        >
          Rating
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
  <div>
    Done
  </div>
  <button>
    <span
      class="text"
    >
      backToTopButton
    </span>
    <i
      aria-hidden="true"
      class="fa fa-arrow-up"
    />
  </button>
</div>
`;

exports[`should render multi page table 1`] = `
<div>
  <div />
  <table>
    <caption>
      Empty
    </caption>
    <thead>
      <tr>
        <th
          scope="col"
        >
          Product
        </th>
        <th
          scope="col"
        >
          Overview
        </th>
        <th
          scope="col"
        >
          Price
        </th>
        <th
          scope="col"
        >
          Rating
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
  <div>
    Done
  </div>
  <button>
    <span
      class="text"
    >
      backToTopButton
    </span>
    <i
      aria-hidden="true"
      class="fa fa-arrow-up"
    />
  </button>
</div>
`;

exports[`should render one page table 1`] = `
<div>
  <div />
  <table>
    <caption>
      One page
    </caption>
    <thead>
      <tr>
        <th
          scope="col"
        >
          Product
        </th>
        <th
          scope="col"
        >
          Overview
        </th>
        <th
          scope="col"
        >
          Price
        </th>
        <th
          scope="col"
        >
          Rating
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          data-label="Product"
        >
          <div
            class="container"
          >
            <div
              class="product"
            >
              <img
                alt="VATTENTÄT"
                class="fit-picture"
                src="http://www.ikea.com/us/en/images/products/vattentat-kettle-black__0169257_PE323125_S4.JPG"
              />
            </div>
          </div>
        </td>
        <td
          data-label="Overview"
        >
          Whistle function when water boils. Made from stainless steel, which is durable and easy to clean.
        </td>
        <td
          data-label="Price"
        >
          12.99
        </td>
        <td
          data-label="Rating"
        >
          3
        </td>
      </tr>
      <tr>
        <td
          data-label="Product"
        >
          <div
            class="container"
          >
            <div
              class="product"
            >
              <img
                alt="SÄTTA"
                class="fit-picture"
                src="http://www.ikea.com/us/en/images/products/satta-knob-assorted-colors__0216460_PE372109_S4.JPG"
              />
            </div>
          </div>
        </td>
        <td
          data-label="Overview"
        />
        <td
          data-label="Price"
        >
          1.99
        </td>
        <td
          data-label="Rating"
        >
          3
        </td>
      </tr>
      <tr>
        <td
          data-label="Product"
        >
          <div
            class="container"
          >
            <div
              class="product"
            >
              <img
                alt="VARNHEM"
                class="fit-picture"
                src="http://www.ikea.com/us/en/images/products/varnhem-knob-gray__0185820_PE337895_S4.JPG"
              />
            </div>
          </div>
        </td>
        <td
          data-label="Overview"
        >
          These nickel-plated knobs are easy to grip and add a traditional look to your kitchen.
        </td>
        <td
          data-label="Price"
        >
          5.99
        </td>
        <td
          data-label="Rating"
        >
          3
        </td>
      </tr>
      <tr>
        <td
          data-label="Product"
        >
          <div
            class="container"
          >
            <div
              class="product"
            >
              <img
                alt="ASPEKT"
                class="fit-picture"
                src="http://www.ikea.com/us/en/images/products/aspekt-knife-sharpener-black__75038_PE192705_S4.JPG"
              />
            </div>
          </div>
        </td>
        <td
          data-label="Overview"
        />
        <td
          data-label="Price"
        >
          4.79
        </td>
        <td
          data-label="Rating"
        >
          3
        </td>
      </tr>
      <tr>
        <td
          data-label="Product"
        >
          <div
            class="container"
          >
            <div
              class="product"
            >
              <img
                alt="ATTEST"
                class="fit-picture"
                src="http://www.ikea.com/us/en/images/products/attest-knob-gray__81916_PE207196_S4.JPG"
              />
            </div>
          </div>
        </td>
        <td
          data-label="Overview"
        />
        <td
          data-label="Price"
        >
          4.99
        </td>
        <td
          data-label="Rating"
        >
          4
        </td>
      </tr>
      <tr>
        <td
          data-label="Product"
        >
          <div
            class="container"
          >
            <div
              class="product"
            >
              <img
                alt="NORRBYN"
                class="fit-picture"
                src="http://www.ikea.com/us/en/images/products/norrbyn-knob-assorted-colors__0256825_PE400851_S4.JPG"
              />
            </div>
          </div>
        </td>
        <td
          data-label="Overview"
        />
        <td
          data-label="Price"
        >
          3.99
        </td>
        <td
          data-label="Rating"
        >
          5
        </td>
      </tr>
      <tr>
        <td
          data-label="Product"
        >
          <div
            class="container"
          >
            <div
              class="product"
            >
              <img
                alt="VÄRDE"
                class="fit-picture"
                src="http://www.ikea.com/us/en/images/products/varde-knob__20420_PE042915_S4.JPG"
              />
            </div>
          </div>
        </td>
        <td
          data-label="Overview"
        />
        <td
          data-label="Price"
        >
          5.99
        </td>
        <td
          data-label="Rating"
        >
          5
        </td>
      </tr>
      <tr>
        <td
          data-label="Product"
        >
          <div
            class="container"
          >
            <div
              class="product"
            >
              <img
                alt="KALKGRUND"
                class="fit-picture"
                src="http://www.ikea.com/us/en/images/products/kalkgrund-knob-hanger__0326134_PE517506_S4.JPG"
              />
            </div>
          </div>
        </td>
        <td
          data-label="Overview"
        >
          No visible screws, as the hardware is concealed. The chrome finish is durable and resistant to corrosion.
        </td>
        <td
          data-label="Price"
        >
          7.99
        </td>
        <td
          data-label="Rating"
        >
          1
        </td>
      </tr>
      <tr>
        <td
          data-label="Product"
        >
          <div
            class="container"
          >
            <div
              class="product"
            >
              <img
                alt="BUDSKÄR"
                class="fit-picture"
                src="http://www.ikea.com/us/en/images/products/budskar-kitchen-faucet__0148875_PE307217_S4.JPG"
              />
            </div>
          </div>
        </td>
        <td
          data-label="Overview"
        >
          10-year Limited Warranty. Read about the terms in the Limited Warranty brochure.
        </td>
        <td
          data-label="Price"
        >
          199
        </td>
        <td
          data-label="Rating"
        >
          3
        </td>
      </tr>
      <tr>
        <td
          data-label="Product"
        >
          <div
            class="container"
          >
            <div
              class="product"
            >
              <img
                alt="ORRNÄS"
                class="fit-picture"
                src="http://www.ikea.com/us/en/images/products/orrnas-knob__0179852_PE332009_S4.JPG"
              />
            </div>
          </div>
        </td>
        <td
          data-label="Overview"
        >
          These brushed stainless-steel knobs give a modern look to your kitchen.
        </td>
        <td
          data-label="Price"
        >
          3.99
        </td>
        <td
          data-label="Rating"
        >
          5
        </td>
      </tr>
      <tr>
        <td
          data-label="Product"
        >
          <div
            class="container"
          >
            <div
              class="product"
            >
              <img
                alt="GOTTLÖSA"
                class="fit-picture"
                src="http://www.ikea.com/us/en/images/products/gottlosa-knob-gray__0185875_PE337938_S4.JPG"
              />
            </div>
          </div>
        </td>
        <td
          data-label="Overview"
        >
          These stainless-steel knobs are softly rounded and add a modern touch to your kitchen.
        </td>
        <td
          data-label="Price"
        >
          5.99
        </td>
        <td
          data-label="Rating"
        >
          3
        </td>
      </tr>
      <tr>
        <td
          data-label="Product"
        >
          <div
            class="container"
          >
            <div
              class="product"
            >
              <img
                alt="LINDRIG"
                class="fit-picture"
                src="http://www.ikea.com/us/en/images/products/lindrig-knife-brown__0278112_PE417596_S4.JPG"
              />
            </div>
          </div>
        </td>
        <td
          data-label="Overview"
        >
          The serrated blade makes it easy to cut meat and ideal to use for eating.
        </td>
        <td
          data-label="Price"
        >
          6.99
        </td>
        <td
          data-label="Rating"
        >
          1
        </td>
      </tr>
      <tr>
        <td
          data-label="Product"
        >
          <div
            class="container"
          >
            <div
              class="product"
            >
              <img
                alt="IKEA PS 2014"
                class="fit-picture"
                src="http://www.ikea.com/us/en/images/products/ikea-ps-knob-assorted-colors__0240271_PE379900_S4.JPG"
              />
            </div>
          </div>
        </td>
        <td
          data-label="Overview"
        >
          Together with the wall rail and the other accessories in the IKEA PS 2014 series you can create different combinations to suit your needs and your home.
        </td>
        <td
          data-label="Price"
        >
          10
        </td>
        <td
          data-label="Rating"
        >
          5
        </td>
      </tr>
      <tr>
        <td
          data-label="Product"
        >
          <div
            class="container"
          >
            <div
              class="product"
            >
              <img
                alt="ELVERDAM"
                class="fit-picture"
                src="http://www.ikea.com/us/en/images/products/elverdam-kitchen-faucet-gray__0173584_PE327697_S4.JPG"
              />
            </div>
          </div>
        </td>
        <td
          data-label="Overview"
        >
          10-year Limited Warranty. Read about the terms in the Limited Warranty brochure.
        </td>
        <td
          data-label="Price"
        >
          149
        </td>
        <td
          data-label="Rating"
        >
          5
        </td>
      </tr>
      <tr>
        <td
          data-label="Product"
        >
          <div
            class="container"
          >
            <div
              class="product"
            >
              <img
                alt="SNITTA"
                class="fit-picture"
                src="http://www.ikea.com/us/en/images/products/snitta-knife-black__0267720_PE407480_S4.JPG"
              />
            </div>
          </div>
        </td>
        <td
          data-label="Overview"
        >
          The serrated blade makes it easy to cut meat and ideal to use for eating.
        </td>
        <td
          data-label="Price"
        >
          3.99
        </td>
        <td
          data-label="Rating"
        >
          1
        </td>
      </tr>
      <tr>
        <td
          data-label="Product"
        >
          <div
            class="container"
          >
            <div
              class="product"
            >
              <img
                alt="IKEA PS 2014"
                class="fit-picture"
                src="http://www.ikea.com/us/en/images/products/ikea-ps-knob-assorted-colors__0240308_PE379913_S4.JPG"
              />
            </div>
          </div>
        </td>
        <td
          data-label="Overview"
        >
          Together with the wall rail and the other accessories in the IKEA PS 2014 series you can create different combinations to suit your needs and your home.
        </td>
        <td
          data-label="Price"
        >
          6
        </td>
        <td
          data-label="Rating"
        >
          5
        </td>
      </tr>
    </tbody>
  </table>
  <div>
    Done
  </div>
  <button>
    <span
      class="text"
    >
      backToTopButton
    </span>
    <i
      aria-hidden="true"
      class="fa fa-arrow-up"
    />
  </button>
</div>
`;
